### ISSUE TITLE
A parsing error

### ISSUE THREAD

**Issue by specke** [z00m128/sjasmplus#138]
| Version 	| Platform 	| Topic 	|
|---------	|----------	|-------	|
| v1.18.1 	| Windows 7 	| conditional assembly |

This error is somewhat difficult to explain. It manifests itself if you have a pair of nested conditional instructions, and the innermost conditional starts with a label. When the outer conditional is disabled, the inner one should be fully silent. However, if the inner conditional starst with a label, it acts as if the outer conditional was not issued at all, and even prints an error message about the closing ENDIF for the outer conditional. The code snippet below summarizes the situation.

```
	DEVICE ZXSPECTRUM48

;	DEFINE BORDER
;	DEFINE RED

	ORG 32768

RUNME:

	IFDEF BORDER

@Border:	IFNDEF RED
			ld a,1 : out (254),a
		ELSE
			ld a,2 : out (254),a
		ENDIF ; RED

	ENDIF

		di : halt

	SAVESNA "quicktest.sna", RUNME
```
For clarification: if you move IFNDEF RED onto the next line from @ Border, the code will work as expected.

**Comment by ped7g** [comment]
The fix is not perfect, works only when `--dirbol` is not used, and doesn't work for temporary labels (the number-ones), but should be good enough for most regular use-cases, like your example.

If anyone hits this same issue with temporary label, or is using `--dirbol`, please split your source into two lines ("won't fix" on sjasmplus side).

The fully correct fix is not trivial, as the sjasmplus is taking big shortcuts to skip the disabled (false condition) block in current architecture, so fully correct fix would have to either duplicate the label parser for disabled block, or change the internal architecture to do disabled blocks differently.

**Comment by specke** [comment]
Many thanks for fixing it so quickly. As for your other comments - the main issue from my perspective was that it took me a couple of days to convince myself that it is actually a bug, and then another couple of days to construct a minimal example. I thought for a while that I must have mismatched some of the IF/ELSE/ENDIF statements (I triggered this bug in a fairly long library file with quite a few conditionals). It is not a very obvious bug, as far as bugs go.

I agree that moving the conditional onto the next line from the label is an easy and mostly satisfactory fix, but how to recognize it as an assembler issue if you have not seen something like this before? Maybe, it could be a good idea to add a comment about this in the documentation?

**Comment by ped7g** [comment]
yeah, generally this is very difficult, even if I would add all those tiny details of internal implementation to docs (and I'm not aware of them usually until someone runs into them :) ), most of the people don't read the docs any way.

What have worked for me since like ~1990 is to let the assembler produce the listing file (if it is capable of, and sjasmplus is), and then check what was assembler producing. If you would try with your example with 1.18.2, you can see how the ELSE branch suddenly belongs to the top-level `IFDEF BORDER` instead of expected `IFNDEF RED` (and then the last IFDEF is orphaned) - which just doesn't make sense, neither common sense or can't be matched with sjasmplus docs. At such point just opening issue here is valid approach. In "worst" case I will show you why the assembler is actually right, but if you did find actual bug, I could confirm it and become aware of it.

Ad docs mention about this particular topic... I don't see many people using temporary labels (they are kinda code-smell IMO, so I'm glad people don't use them much :) ), and similarly people requiring `--dirbol` have to deal with many more ambiguous situations, so I guess they are used to that. To cover "common" usage like you posted, that was possible to resolve with 4 extra lines, and ~1% slow down in assembling of disabled blocks, so that's acceptable for me.

>  that it took me a couple of days to convince myself that it is actually a bug

I will take this personally, as compliment to my last two years crusade to make sjasmplus bugfixed and reliable enough. :) If you would start with 1.10.x (my first fixes went into 1.11.0), you would run into weird bugs probably lot more sooner.

But the fact of life is, that sjasmplus internal implementation is not the prettiest and most clever piece of code, and I did spend those two years mostly by adding test coverage and just bug-fixing it, rarely refactoring, and even the refactored parts are probably not the most elegant and prettiest pieces of code in the world -> so it's not *that* shocking you did hit a bug in sjasmplus. I hope this will be one of those very rare occasions, and overall it will work for you well and reliably.

Sorry for the inconvenience.
